### PR TITLE
feat: Add data availability validation for Celestia DA #422

### PR DESCRIPTION
**Pull request by srene**
# PR Standards

## Opening a pull request should be able to meet the following requirements

---

For Author:

- [X]  Targeted PR against correct branch
- [X]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [X]  Linked to Github issue with discussion and accepted design
- [X]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: da/celestia/celestia.go [hunk 1]
```diff
@@ -223,36 +232,105 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
...
+			}
+		}
+	}
+
+}
+
+func (c *DataAvailabilityLayerClient) RetrieveBatches(daMetaData *da.DAMetaData) da.ResultRetrieveBatch {
+
+	//Just for backward compatibility, in case no commitments are sent from the Hub, batch can be retrieved using previous implementation.
+	if daMetaData.Commitments == nil || len(daMetaData.Commitments) == 0 {
```
Comment (omritoptix) [inline da/celestia/celestia.go:269]: This is actually a security hole so not sure we can release it like that.
The big majority of Froopyland don't run full nodes. Once we upgrade them assuming not synced, they will have to sync from a snapshot from the point that state updates must contain commitments.
I suggest removing it.

File: da/celestia/celestia.go [hunk 2]
```diff
@@ -223,36 +232,105 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
...
+							Code:    da.StatusBlobNotFound,
+							Message: "Blob not found",
+						},
+					}
+				}
+
+				var batch pb.Batch
+				err = proto.Unmarshal(blob.Data, &batch)
+				if err != nil {
+					c.logger.Error("failed to unmarshal block", "daHeight", daMetaData.Height, "error", err)
```
Comment (omritoptix) [inline da/celestia/celestia.go:303]: failed to unmarshal `batch` instead of `block`

File: block/retriever.go [hunk 3]
```diff
@@ -99,19 +99,54 @@ func (m *Manager) processNextDABatch(ctx context.Context, daHeight uint64) error
 	return nil
 }
 
-func (m *Manager) fetchBatch(daHeight uint64) (da.ResultRetrieveBatch, error) {
+func (m *Manager) fetchBatch(daMetaData *da.DAMetaData) (da.ResultRetrieveBatch, error) {
 	var err error
-	batchRes := m.retriever.RetrieveBatches(daHeight)
+	availRes := m.retriever.CheckBatchAvailability(daMetaData)
+
+	switch availRes.Code {
```
Comment (omritoptix) [inline block/retriever.go:106]: imo leaving the code each status should indiciate a different handling. i.e `success` means on thing, `errorGettingProof` can be another and `ProofInvalid` can be another one but if they have the same handling route than the status code a bit looses the point IMO. If you want to add extra info to each status you can add it in the `message` field.

File: block/retriever.go [hunk 4]
```diff
@@ -99,19 +99,54 @@ func (m *Manager) processNextDABatch(ctx context.Context, daHeight uint64) error
...
+		err = fmt.Errorf("Unable to get proof on height %d for the commitment", daMetaData.Height)
+
+	case da.StatusProofNotMatching:
+		//The proofs are obtained for the commitment, but not matching with the span (index, length) committed to the Hub.
+		err = fmt.Errorf("Span not matching the commitment")
+	case da.StatusError:
+		//There's been an issue validating proofs
+		err = fmt.Errorf("Error validating batch")
+	}
+	if !availRes.DataAvailable {
```
Comment (omritoptix) [inline block/retriever.go:119]: Also here would change to `availabilityRes` as we have `avail` and it's confusing

File: block/retriever.go [hunk 5]
```diff
@@ -99,19 +99,54 @@ func (m *Manager) processNextDABatch(ctx context.Context, daHeight uint64) error
...
+	}
+	//batchRes.MetaData includes proofs necessary to open disputes with the Hub
+	batchRes := m.retriever.RetrieveBatches(daMetaData)
+	batchRes.MetaData = availRes.MetaData
+
 	switch batchRes.Code {
 	case da.StatusError:
-		err = fmt.Errorf("failed to retrieve batch from height %d: %s", daHeight, batchRes.Message)
+		err = fmt.Errorf("failed to retrieve batch from height %d: %s", daMetaData.Height, batchRes.Message)
 	case da.StatusTimeout:
```
Comment (omritoptix) [inline block/retriever.go:136]: Same here with the switch case of enums like above

File: block/retriever.go [hunk 6]
```diff
@@ -99,19 +99,54 @@ func (m *Manager) processNextDABatch(ctx context.Context, daHeight uint64) error
...
 	var err error
-	batchRes := m.retriever.RetrieveBatches(daHeight)
+	availRes := m.retriever.CheckBatchAvailability(daMetaData)
+
+	switch availRes.Code {
+	case da.StatusUnableToGetProof:
+		//Is not possible to obtain proofs for the specific commitment in the specific height.
+		//This means there is no matching blob for that commitment
+		err = fmt.Errorf("Unable to get proof on height %d for the commitment", daMetaData.Height)
+
```
Comment (omritoptix) [inline block/retriever.go:111]: no need for those spaces..

File: block/retriever.go (hunk 4 above)
Comment (omritoptix) [reply block/retriever.go:119]: Also seems like `DataAvailable`==`statusSuccess` as from it's description. so not sure why both are needed.

File: block/retriever.go [hunk 7]
```diff
@@ -99,19 +99,54 @@ func (m *Manager) processNextDABatch(ctx context.Context, daHeight uint64) error
...
+		//The proofs are obtained for the commitment, but not matching with the span (index, length) committed to the Hub.
+		err = fmt.Errorf("Span not matching the commitment")
+	case da.StatusError:
+		//There's been an issue validating proofs
+		err = fmt.Errorf("Error validating batch")
+	}
+	if !availRes.DataAvailable {
+		err = fmt.Errorf("no batches found on height %d", daMetaData.Height)
+		//There is no point on fetching the data
+		batchRes := da.ResultRetrieveBatch{}
```
Comment (omritoptix) [inline block/retriever.go:122]: it's better to just insert the fields in the ResultRetrieveBatch while you create it vs adding it after

File: block/retriever.go [hunk 8]
```diff
@@ -99,19 +99,54 @@ func (m *Manager) processNextDABatch(ctx context.Context, daHeight uint64) error
...
+		batchRes.Code = da.StatusError
+		batchRes.Message = "Error validating data"
+		batchRes.MetaData = availRes.MetaData
+		return batchRes, err
```
Comment (omritoptix) [inline block/retriever.go:126]: I guess if data is not available we need to seperate to 2 cases:
1. not available due to RPC down etc - in this case - should be retried
2. not available due to fraud, which is another case.

Lets not handle this now but lets make sure we have an issue to handle this.

File: block/retriever.go [hunk 9]
```diff
@@ -99,19 +99,54 @@ func (m *Manager) processNextDABatch(ctx context.Context, daHeight uint64) error
...
+
+	}
+	//batchRes.MetaData includes proofs necessary to open disputes with the Hub
+	batchRes := m.retriever.RetrieveBatches(daMetaData)
+	batchRes.MetaData = availRes.MetaData
```
Comment (omritoptix) [inline block/retriever.go:131]: error should be checked before assignment

File: block/retriever.go [hunk 10]
```diff
@@ -99,19 +99,54 @@ func (m *Manager) processNextDABatch(ctx context.Context, daHeight uint64) error
...
+
 	switch batchRes.Code {
 	case da.StatusError:
-		err = fmt.Errorf("failed to retrieve batch from height %d: %s", daHeight, batchRes.Message)
+		err = fmt.Errorf("failed to retrieve batch from height %d: %s", daMetaData.Height, batchRes.Message)
```
Comment (omritoptix) [inline block/retriever.go:135]: creation of error based on case is a bit dirty IMO. As I see it:
1. we can either return the original Error
2. we can simply log the message and return original error
3. we can have the `message` field holding the error which can be either a string or an error object which we call Error() on.

File: settlement/dymension/dymension.go [hunk 11]
```diff
@@ -407,8 +407,14 @@ func (d *HubClient) convertBatchToMsgUpdateState(batch *types.Batch, daClient da
 		return nil, err
 	}
 
-	DAMetaData := &settlement.DAMetaData{
-		Height: daResult.DAHeight,
+	height := uint64(0)
```
Comment (omritoptix) [inline settlement/dymension/dymension.go:410]: why is this necessary?

File: settlement/dymension/dymension.go [hunk 12]
```diff
@@ -407,8 +407,14 @@ func (d *HubClient) convertBatchToMsgUpdateState(batch *types.Batch, daClient da
...
+	if daResult.MetaData != nil {
+		height = daResult.MetaData.Height
```
Comment (omritoptix) [inline settlement/dymension/dymension.go:412]: why should the daResult.MetData be nil? why are we checking this?

File: settlement/dymension/dymension.go [hunk 13]
```diff
@@ -407,8 +407,14 @@ func (d *HubClient) convertBatchToMsgUpdateState(batch *types.Batch, daClient da
...
+	}
+
+	//TODO (srene): Include extra dataMetaDATA to the Hub for inclusion validation
+	DAMetaData := &da.DAMetaData{
+		Height: height,
```
Comment (omritoptix) [inline settlement/dymension/dymension.go:417]: Why are we jus not setting the height to the daResult.Height right away like before?

File: da/celestia/celestia.go [hunk 14]
```diff
@@ -180,7 +186,11 @@ func (c *DataAvailabilityLayerClient) GetClientType() da.Client {
 
 // SubmitBatch submits a batch to the DA layer.
 func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultSubmitBatch {
+
+	var blobs [][]byte
+
 	data, err := batch.MarshalBinary()
+	blobs = [][]byte{data}
```
[the remark this answers is not in the document]
Comment (srene) [reply da/celestia/celestia.go:193]: sure. moved

File: da/celestia/celestia.go [hunk 15]
```diff
@@ -189,30 +199,29 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
...
 	}
-
-	blockBlob, err := blob.NewBlobV0(c.config.NamespaceID.Bytes(), data)
-	if err != nil {
+	if len(data) > DefaultMaxBytes {
 		return da.ResultSubmitBatch{
 			BaseResult: da.BaseResult{
 				Code:    da.StatusError,
-				Message: err.Error(),
+				Message: "Batch size bigger than maximum blob size",
```
[the remark this answers is not in the document]
Comment (srene) [reply da/celestia/celestia.go:206]: ok

File: da/celestia/celestia.go [hunk 16]
```diff
@@ -189,30 +199,29 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
...
 
 	for {
 		select {
 		case <-c.ctx.Done():
 			c.logger.Debug("Context cancelled")
 			return da.ResultSubmitBatch{}
 		default:
-			txResponse, err := c.rpc.SubmitPayForBlob(c.ctx, math.NewInt(fees), gasWanted, blobs)
+			gasPrice := float64(-1)
+			if c.config.GasPrices >= 0 {
```
[the remark this answers is not in the document]
Comment (srene) [reply da/celestia/celestia.go:218]: true. moved

[the remark this answers is not in the document]
Comment (srene) [reply da/celestia/celestia.go:218]: I think I got it from rollkit example to indicate is not set, but I actually removed gasPrices as input parameter for the submit() function, because it can be get later on from the config anyway, and no need to pass as a parameter.

File: da/celestia/celestia.go [hunk 17]
```diff
@@ -189,30 +199,29 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
...
+				gasPrice = c.config.GasPrices
+			}
+			//TODO(srene):  Split batch in multiple blobs if necessary if supported
+			height, commitments, blobs, err := c.submit(blobs, gasPrice)
+
+			c.logger.Info("DA batch submitted", "commitments", len(commitments), "blobs", len(blobs))
```
[the remark this answers is not in the document]
Comment (srene) [reply da/celestia/celestia.go:224]: moved

[the remark this answers is not in the document]
Comment (srene) [reply da/celestia/celestia.go:224]: I also removed returned blobs from submit function, since it was only used for logging

File: block/retriever.go (hunk 9 above)
Comment (srene) [reply block/retriever.go:131]: done

File: settlement/dymension/dymension.go (hunk 13 above)
Comment (srene) [reply settlement/dymension/dymension.go:417]: this was to avoid null pointer errors caused by a test returning empty batch result, which should not happen anyway because it will panic before. I adapted the test and removed this

File: da/avail/avail.go [hunk 18]
```diff
@@ -155,11 +155,12 @@ func (c *DataAvailabilityLayerClient) GetClientType() da.Client {
 }
 
 // RetrieveBatch retrieves batch from DataAvailabilityLayerClient instance.
-func (c *DataAvailabilityLayerClient) RetrieveBatches(dataLayerHeight uint64) da.ResultRetrieveBatch {
-	blockHash, err := c.client.GetBlockHash(dataLayerHeight)
+func (c *DataAvailabilityLayerClient) RetrieveBatches(daMetaData *da.DASubmitMetaData) da.ResultRetrieveBatch {
```
Comment (mtsitrin) [inline da/avail/avail.go:158]: something off with the naming, we pass `DASubmitMetaData` to *Retrieve* function

File: da/avail/avail.go [hunk 19]
```diff
@@ -398,9 +402,9 @@ func (c *DataAvailabilityLayerClient) broadcastTx(tx []byte) (uint64, error) {
 }
 
 // CheckBatchAvailability checks batch availability in DataAvailabilityLayerClient instance.
-func (c *DataAvailabilityLayerClient) CheckBatchAvailability(dataLayerHeight uint64) da.ResultCheckBatch {
+func (c *DataAvailabilityLayerClient) CheckBatchAvailability(daMetaData *da.DASubmitMetaData) da.ResultCheckBatch {
 	return da.ResultCheckBatch{
-		BaseResult: da.BaseResult{
+		BaseDACheckResult: da.BaseDACheckResult{
```
Comment (mtsitrin) [inline da/avail/avail.go:407]: doesn't it break all avail rollapps? they won't be able to retrieve blocks optimisticlly

File: da/avail/avail.go (hunk 18 above)
Comment (srene) [reply da/avail/avail.go:158]: I named DASubmitMetadata, the DA metadata result of the submission, and used for retrieving the data.
I named DACheckMetaData, the DA metadata resulting of the availability checks including inclusion proofs for disputes. 
I can change the naming if you think is not clear enough

File: da/celestia/celestia.go [hunk 20]
```diff
@@ -189,30 +196,29 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
 			},
 		}
 	}
+	blobs = [][]byte{data}
 
-	blockBlob, err := blob.NewBlobV0(c.config.NamespaceID.Bytes(), data)
-	if err != nil {
+	if len(data) > celtypes.DefaultMaxBytes {
```
Comment (mtsitrin) [inline da/celestia/celestia.go:201]: this value should be validated for dymint config as well (we have max bytes for batch there)

File: da/celestia/celestia.go [hunk 21]
```diff
@@ -189,30 +196,29 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
...
 
 	for {
 		select {
 		case <-c.ctx.Done():
 			c.logger.Debug("Context cancelled")
 			return da.ResultSubmitBatch{}
 		default:
-			txResponse, err := c.rpc.SubmitPayForBlob(c.ctx, math.NewInt(fees), gasWanted, blobs)
+
+			c.logger.Info("Submitting DA batch")
```
Comment (mtsitrin) [inline da/celestia/celestia.go:217]: log this above, as it will be printed for each retry otherwise

File: da/celestia/celestia.go [hunk 22]
```diff
@@ -180,6 +184,9 @@ func (c *DataAvailabilityLayerClient) GetClientType() da.Client {
 
 // SubmitBatch submits a batch to the DA layer.
 func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultSubmitBatch {
+
+	var blobs [][]byte
```
Comment (mtsitrin) [inline da/celestia/celestia.go:188]: redundent. aleardy initilized by `blobs = [][]byte{data}`

File: da/celestia/celestia.go [hunk 23]
```diff
@@ -223,40 +229,106 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
...
-			if txResponse == nil {
-				err := errors.New("txResponse is nil")
-				c.logger.Error("Failed to submit DA batch", "error", err)
+			daMetaData := &da.DASubmitMetaData{
+				Height:      height,
+				Commitments: commitments,
+			}
+
+			availabilityResult := c.CheckBatchAvailability(daMetaData)
+			if availabilityResult.Code != da.StatusSuccess || !availabilityResult.DataAvailable {
```
Comment (mtsitrin) [inline da/celestia/celestia.go:238]: `availabilityResult.Code != da.StatusSuccess`
will fail for avail

File: da/celestia/celestia.go [hunk 24]
```diff
@@ -285,10 +357,264 @@ func (c *DataAvailabilityLayerClient) RetrieveBatches(dataLayerHeight uint64) da
...
+	defer cancel()
+	isIncluded, error := c.rpc.Included(ctx, height, c.config.NamespaceID.Bytes(), proof, commitment)
+	return isIncluded, error
+}
+
+func (c *DataAvailabilityLayerClient) getDataAvailabilityHeaders(height uint64) (*header.DataAvailabilityHeader, error) {
+
+	c.logger.Info("Getting Celestia extended headers via RPC call")
+	ctx, cancel := context.WithTimeout(c.ctx, c.txPollingRetryDelay)
+	defer cancel()
```
Comment (mtsitrin) [inline da/celestia/celestia.go:573]: no need for it

File: da/celestia/celestia.go [hunk 25]
```diff
@@ -285,10 +357,264 @@ func (c *DataAvailabilityLayerClient) RetrieveBatches(dataLayerHeight uint64) da
...
+		commitments = append(commitments, commitment)
+	}
+	return blobs, commitments, nil
+}
+
+func (c *DataAvailabilityLayerClient) validateProof(height uint64, commitment da.Commitment, proof *blob.Proof) (bool, error) {
+
+	c.logger.Info("Getting inclusion validation via RPC call")
+	ctx, cancel := context.WithTimeout(c.ctx, c.txPollingRetryDelay)
+	defer cancel()
```
Comment (mtsitrin) [inline da/celestia/celestia.go:564]: no need for it

File: da/celestia/celestia.go [hunk 26]
```diff
@@ -285,10 +357,264 @@ func (c *DataAvailabilityLayerClient) RetrieveBatches(dataLayerHeight uint64) da
...
+				CheckMetaData: DACheckMetaData,
+			},
+		}
+	}
+	DACheckMetaData.Root = dah.Hash()
+	for i, commitment := range daMetaData.Commitments {
+		included := false
+
+		proof, err := c.getProof(daMetaData.Height, commitment)
+		if err != nil || proof == nil {
```
Comment (mtsitrin) [inline da/celestia/celestia.go:394]: err valiue not needed? not logged

File: da/celestia/celestia.go [hunk 27]
```diff
@@ -285,10 +357,264 @@ func (c *DataAvailabilityLayerClient) RetrieveBatches(dataLayerHeight uint64) da
...
+		for i, proof := range nmtProofs {
+			if i == 0 {
+				indexes = append(indexes, proof.Start())
+			}
+			shares += proof.End() - proof.Start()
+		}
+		numShares = append(numShares, shares)
+
+		if daMetaData.Indexes != nil && daMetaData.Lengths != nil {
+			if indexes[i] != daMetaData.Indexes[i] || shares != daMetaData.Lengths[i] {
```
Comment (mtsitrin) [inline da/celestia/celestia.go:421]: how u know that indexes[i] exists?
indexes appended once, i runs over nmtProofs

Comment (srene) [reply da/celestia/celestia.go:421]: if there is no index is because there is no proof returned in previous step, and therefore the function exited before.

File: da/celestia/celestia.go (hunk 24 above)
Comment (srene) [reply da/celestia/celestia.go:573]: it's there to force  a timeout in case rpc call does not timeout by itself (which was happening for some problems in celestia)

File: da/celestia/celestia.go (hunk 25 above)
Comment (srene) [reply da/celestia/celestia.go:564]: same here

File: da/celestia/celestia.go [hunk 28]
```diff
@@ -223,40 +229,106 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
...
+			}
+
+			availabilityResult := c.CheckBatchAvailability(daMetaData)
+			if availabilityResult.Code != da.StatusSuccess || !availabilityResult.DataAvailable {
+				c.logger.Error("Unable to confirm submitted blob availability. Retrying")
 				res, err := da.SubmitBatchHealthEventHelper(c.pubsubServer, c.ctx, false, err)
 				if err != nil {
 					return res
 				}
 				time.Sleep(c.submitRetryDelay)
```
[the remark this answers is not in the document]
Comment (srene) [reply da/celestia/celestia.go:244]: this shouldn't happen, because if the submission was successful, there should be valid proofs available for the blob. however, in case it happens, it makes sense to me to resubmit again, disregarding previous submission result, because from rpc calls it is not possible to know whether there was a problem in the submitted blob or in getting the proofs.

File: da/celestia/celestia.go [hunk 29]
```diff
@@ -285,10 +357,264 @@ func (c *DataAvailabilityLayerClient) RetrieveBatches(dataLayerHeight uint64) da
...
+// Submit submits the Blobs to Data Availability layer.
+func (c *DataAvailabilityLayerClient) submit(daBlobs []da.Blob) (uint64, []da.Commitment, error) {
+	blobs, commitments, err := c.blobsAndCommitments(daBlobs)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	options := openrpc.DefaultSubmitOptions()
+
+	if c.config.GasPrices >= 0 {
```
[the remark this answers is not in the document]
Comment (srene) [reply da/celestia/celestia.go:500]: ok. check removed

File: da/celestia/celestia.go (hunk 22 above)
Comment (srene) [reply da/celestia/celestia.go:188]: ok. removed

**Review by omritoptix — APPROVED** [review]
